web/app/js/components/Top.jsx

```diff
@@ -0,0 +1,361 @@
+  }
+
+  constructor(props) {
+    super(props);
+    this.api = this.props.api;
+    this.loadFromServer = this.loadFromServer.bind(this);
+
+    this.state = {
+      tapResultsById: {},
+      topEventIndex: {},
+      error: null,
+      resourcesByNs: {},
+      authoritiesByNs: {},
+      query: {
+        resource: "",
+        namespace: "",
+        toResource: "",
+        toNamespace: "",
+        method: "",
+        path: "",
+        scheme: "",
+        authority: "",
+        maxRps: defaultMaxRps
+      },
+      maxRowsToStore: 40,
+      awaitingWebSocketConnection: false,
+      tapRequestInProgress: false,
+      pollingInterval: 10000,
+      pendingRequests: false
+    };
+  }
+
+  componentDidMount() {
+    this.startServerPolling();
+  }
+
+  componentWillUnmount() {
+    if (this.ws) {
+      this.ws.close(1000);
+    }
+    this.stopTapStreaming();
+    this.stopServerPolling();
+  }
+
+  onWebsocketOpen = () => {
+    let query = _.cloneDeep(this.state.query);
+    query.maxRps = parseFloat(query.maxRps);
+
+    this.ws.send(JSON.stringify({
+      id: "tap-web",
+      ...query
+    }));
+    this.setState({
+      awaitingWebSocketConnection: false,
+      error: null
+    });
+  }
+
+  onWebsocketRecv = e => {
+    this.indexTapResult(e.data);
+  }
+
+  onWebsocketClose = e => {
+    this.stopTapStreaming();
+
+    if (!e.wasClean) {
+      this.setState({
+        error: {
+          error: `Websocket [${e.code}] ${e.reason}`
+        }
+      });
+    }
+  }
+
+  onWebsocketError = e => {
+    this.setState({
+      error: { error: e.message }
+    });
+
+    this.stopTapStreaming();
+  }
+
+  getResourcesByNs(rsp) {
+    let statTables = _.get(rsp, [0, "ok", "statTables"]);
+    let authoritiesByNs = {};
+    let resourcesByNs = _.reduce(statTables, (mem, table) => {
+      _.map(table.podGroup.rows, row => {
+        if (!mem[row.resource.namespace]) {
+          mem[row.resource.namespace] = [];
+          authoritiesByNs[row.resource.namespace] = [];
+        }
+
+        switch (row.resource.type.toLowerCase()) {
+          case "service":
+            break;
+          case "authority":
+            authoritiesByNs[row.resource.namespace].push(row.resource.name);
+            break;
+          default:
+            mem[row.resource.namespace].push(`${row.resource.type}/${row.resource.name}`);
+        }
+      });
+      return mem;
+    }, {});
+    return {
+      authoritiesByNs,
+      resourcesByNs
+    };
+  }
+
+  parseTapResult = data => {
+    let d = processTapEvent(data);
+
+    if (d.eventType === "rsp") {
+      d.success = parseInt(d.http.responseInit.httpStatus, 10) < 500;
+    } else if (d.eventType === "end") {
+      d.latency = parseFloat(d.http.responseEnd.sinceRequestInit.replace("s", ""));
+      d.completed = true;
+    }
+
+    return d;
+  }
+
+  topEventKey = event => {
+    return [event.source.str, event.destination.str, event.http.requestInit.path].join("_");
+  }
```

Comment:
Just noting that this key function differs from what @adleong is using in #1435. His approach is to use the event's ID field, but that's unfortunately not guaranteed to be unique due to #1456. Given that, this approach looks good to me, but we should definitely standardize the CLI and web UI sooner rather than later.

Reply:
If I'm reading this correctly, this actually does mirror what I had in the CLI.  There are two different aggregation steps.  First, we receive tap events and aggregate them by id until we've collected the reqInit, rspInit, and rspEnd.  Once we have all 3 parts of the request, we send it over to the table where whole requests are aggregated by (src, dst, path).

In order to avoid relying on event id, I've now changed the CLI implementation to do that first aggregation step by (src, dst, stream).  That same change should probably made here as well for the same reason.

Reply:
yeah, that's correct, I am currently using `id.base:id.stream` to collate the tap results, and using `(src,dst,path)` to key the top results. I'll change this to `(src,dst,stream)`

Reply:
Per in-person conversation, I think this key func needs to strip ports from `source.str` and `destination.str` to match how the CLI is handling it in #1435.

Reply:
Nevermind -- looks like you're already stripping the port when those strings are created.